Paul Hands Legal Agreement - Master - Isla - Added more elections to the isla GMSLA namespace

regnosys-prod-user opened the pull request.

paulhands committed:
Legal Agreement - Master - Isla - Added more elections to the isla GMSLA namespace

regnosys-prod-user commented:
# *Legal Agreement - Master - Isla - Added more elections to the isla GMSLA namespace*

_Background_

Now that Foundations is merged, it is a good time to start adding in some (not all) of the extra elections that will help describe a Legal Agreement "GMSLA" in more detail in the CDM representative space. This should be iterated over, and slowly build to a full picture of representing this type of contract.

_What is being released?_

Added more elections to the isla namespace, with enumerated values. I propose we keep these distinct and separate to begin with, so each agreement can grow organically. At a time in the future when definitions that overlap can be spotted, we can revisit to make shared types and enums, as necessary

_Review Directions_

Changes can be reviewed in PR: [#3968](https://github.com/finos/common-domain-model/pull/3968)

> [!NOTE]
> This comment was generated via [Rosetta](https://rosetta-technology.io/).

paulhands commented:
Note this applies to https://github.com/finos/common-domain-model/issues/3967

chrisisla reviewed (commented):
Hi @paulhands 
I've added a few minor comments but it is really good to see the skeleton introduced earlier in the year to support strongly typed master agreements being put into practice 🙂  
Thanks!

paulhands committed:
Updated the Automatic Early termination piece to more coincide with the ISLA clause library

regnosys-prod-user commented:
_What is being released?_

Updated the Automatic Early termination piece to more coincide with the ISLA clause library

> [!NOTE]
> This comment was generated via [Rosetta](https://rosetta-technology.io/).

paulhands committed:
Updated the paragraph 54 elections aggregation to more accurately reflect the clause library Note...

paulhands commented on rosetta-source/src/main/rosetta/legaldocumentation-master-isla-enum.rosetta:
```diff
@@ -0,0 +1,19 @@
+namespace cdm.legaldocumentation.master.isla
+version "${project.version}"
+
+enum ApplicabilityOfGMSLAElectionEnum: <"The potential options for specifying whether a particular paragraph applies in a GMSLA">
```
GMSLA remove

paulhands committed:
Addressed comments from Chris Rayner Removed the GMSLA from the ApplicabilityOfElectionEnum as th...

regnosys-prod-user commented:
_What is being released?_

Addressed comments from Chris Rayner: - Removed the GMSLA from the ApplicabilityOfElectionEnum, as this will probably move into a shared domain once initial mapping for all master agreements is completed - Updated a Party type to be a CounterPartyRoleEnum instead, as it will always be one of Party 1 or Party 2 - Fixed a validation error that was checking standardDefinitionOfMarketValueUsed being true instead of false.

> [!NOTE]
> This comment was generated via [Rosetta](https://rosetta-technology.io/).

paulhands commented:
@chrisisla have made those changes you suggested, thanks for taking a look at the PR

chrisisla commented on rosetta-source/src/main/rosetta/legaldocumentation-master-isla-type.rosetta:
```diff
@@ -2,5 +2,74 @@ namespace cdm.legaldocumentation.master.isla
+    standardDefinitionOfMarketValueUsed boolean (1..1) <"Is the standard definition of 'Market Value' used in the GMSLA?">
+    definitionOfMarketValueNonStandard string (0..1) <"If the definition of 'Market Value' is non standard, it can be specified here.">
+    aggregation Aggregation (1..1) <"How does paragraph 5.4 (aggregation) apply to the GMSLA?">
+    paragraph5Point6Applies ApplicabilityOfElectionEnum (1..1) <"Does paragraph 5.6 (netting of obligations to deliver Collateral and redeliver Equivalent Collateral) apply to the GMSLA?">
```
Rather than use the paragraph in the name could we use the purpose of the paragraph instead? 
So instead of `paragraph5point6Applies` would it work if we had something like `nettingObligationsApply`? The comment has the paragraph details and the full description too which works nicely.

chrisisla commented on rosetta-source/src/main/rosetta/legaldocumentation-master-isla-type.rosetta:
```diff
@@ -2,5 +2,74 @@ namespace cdm.legaldocumentation.master.isla
+    definitionOfMarketValueNonStandard string (0..1) <"If the definition of 'Market Value' is non standard, it can be specified here.">
+    aggregation Aggregation (1..1) <"How does paragraph 5.4 (aggregation) apply to the GMSLA?">
+    paragraph5Point6Applies ApplicabilityOfElectionEnum (1..1) <"Does paragraph 5.6 (netting of obligations to deliver Collateral and redeliver Equivalent Collateral) apply to the GMSLA?">
+    paragraph6Point4Applies ApplicabilityOfElectionEnum (1..1) <"Does paragraph 6.4 (indemnity for failure to redeliver Equivalent Non-Cash Collateral) apply to the GMSLA?">
```
Same for this attribute, maybe something like `failureIndemnityApplies`, or `failToDeliverIndemnityApplies`?

chrisisla commented on rosetta-source/src/main/rosetta/legaldocumentation-master-isla-type.rosetta:
```diff
@@ -2,5 +2,74 @@ namespace cdm.legaldocumentation.master.isla
+    automaticEarlyTerminationInModifiedForm boolean (1..1) <"Automatic early termination applies in a modified form only">
+    automaticEarlyTerminationOther string (0..1) <"Optional string to capture any further information regarding the automatic early termination even of default.">
+
+    condition AetMustBeApplicableCondition: <"A validation rule to ensure that if aet applies in a modified form or separately for each principle that the applicability is set to be 'applicable'">
```
consider renaming to `AutomaticEarlyTerminationMustBeApplicable` - that removes the AET acronym and we don't need the word "condition" on the end 🙂

chrisisla commented on rosetta-source/src/main/rosetta/legaldocumentation-master-isla-type.rosetta:
```diff
@@ -2,5 +2,74 @@ namespace cdm.legaldocumentation.master.isla
+
+    condition AetMustBeApplicableCondition: <"A validation rule to ensure that if aet applies in a modified form or separately for each principle that the applicability is set to be 'applicable'">
+        if automaticEarlyTerminationInModifiedForm or automaticEarlyTerminationSpecifiedSeparatelyForEachPrinciple
+        then automaticEarlyTerminationApplies = ApplicabilityOfElectionEnum -> Applicable
```
you don't need to specify the enum anymore, so this can be written as:

`then automaticEarlyTerminationApplies = Applicable`

chrisisla commented on rosetta-source/src/main/rosetta/legaldocumentation-master-isla-type.rosetta:
```diff
@@ -2,5 +2,74 @@ namespace cdm.legaldocumentation.master.isla
+    partyElection EventsOfDefaultElection (2..2) <"The specified events of default elections for the respective parties signing the GMSLA.">
+
+type DefaultInterest: <"The definitions regarding default interest relating to the GMSLA.">
+    defaultInterestForParagraph11Point7 DefaultInterestRateSpecificationEnum (1..1) <"How is the default interest rate specified for the purposes of Paragraph 11.7?">
```
Let's have a chat about how to name this rather than use paragraph 11.7

chrisisla commented on rosetta-source/src/main/rosetta/legaldocumentation-master-isla-type.rosetta:
```diff
@@ -2,5 +2,74 @@ namespace cdm.legaldocumentation.master.isla
+        then defaultInterestRateIfOther exists
+
+type ExistingLoans: <"The definitions for existing loans and the applicability of Paragraph 27.4 (only relevant to 2010 GMSLAs).">
+    paragraph27Point4Applies ApplicabilityOfElectionEnum (1..1) <"Does paragraph 27.4 apply?">
```
I believe that paragraph 27.4 is stating that all existing loans will be covered by the agreement in the event of default. Let's have a chat about how to implement this clause.

chrisisla commented on rosetta-source/src/main/rosetta/legaldocumentation-master-isla-type.rosetta:
```diff
@@ -2,5 +2,74 @@ namespace cdm.legaldocumentation.master.isla
+    paragraph27Point4Applies ApplicabilityOfElectionEnum (1..1) <"Does paragraph 27.4 apply?">
+
+type Automation: <"The definitions for automation and the applicability of Paragraph 27.5 (only relevant to 2010 GMSLAs).">
+    paragraph27Point5Applies ApplicabilityOfElectionEnum (1..1) <"Does paragraph 27.5 apply?">
```
Same for this one, paragraph 27.5 is stating that each party can use a third party to automate the processing of loans. So let's see if we can come up with a better name.

chrisisla reviewed (changes requested):
Looking really good, a couple of minor points and it would be good for us to catch up on the names of some of the attributes to see if we can make them more user friendly and less associated to specific paragraphs in the GMSLA 🙂

paulhands committed:
Added missing punctuation to comments Revised the conditionals to remove the ENUM gt value concep...

regnosys-prod-user commented:
_What is being released?_

- Added missing punctuation to comments. - Revised the conditionals to remove the ENUM -&gt; value concept to simply look at the value - Renamed some variables to remove the concept of specific paragraphs in a GMSLA (some more of these need discussing and finalising)

> [!NOTE]
> This comment was generated via [Rosetta](https://rosetta-technology.io/).

paulhands committed:
Further comment grammatical changes

regnosys-prod-user commented:
_What is being released?_

Further comment grammatical changes

> [!NOTE]
> This comment was generated via [Rosetta](https://rosetta-technology.io/).

paulhands committed:
Removed the Default interest concept on the basis that this needs more revision Changed the namin...

regnosys-prod-user commented:
_What is being released?_

Removed the Default interest concept, on the basis that this needs more revision. Changed the naming convention to remove reference to specific paragraphs.

> [!NOTE]
> This comment was generated via [Rosetta](https://rosetta-technology.io/).

chrisisla reviewed (dismissed):
Thanks for making the changes Paul, I think this looks good as a first contribution of a few clauses in the GMSLA.

paulhands committed:
Modified Collateral to be CollateralSpecification to help with type issues on compile

regnosys-prod-user commented:
_What is being released?_

Modified Collateral to be CollateralSpecification to help with type issues on compile

> [!NOTE]
> This comment was generated via [Rosetta](https://rosetta-technology.io/).
